### PR TITLE
Introduce Stateful Callbacks

### LINKED ISSUE #28544: Early stopping patience does not work when resuming from checkpoint
### System Info

- `transformers` version: 4.33.3
- Platform: Linux-4.18.0-348.23.1.el8_5.x86_64-x86_64-with-glibc2.28
- Python version: 3.9.16
- Huggingface_hub version: 0.16.2
- Safetensors version: 0.3.1
- Accelerate version: 0.24.1
- Accelerate config: 	not found
- PyTorch version (GPU?): 2.0.0.post101 (False)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using GPU in script?: Yes
- Using distributed or parallel set-up in script?: No

### Who can help?

@muellerzr and @pacman100

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [X] My own task or dataset (give details below)

### Reproduction

Fundamentally the issue is that the `early_stopping_patience_counter` is not persisted when checkpointing. Consequently, it is [always (re)set to 0](https://github.com/huggingface/transformers/blob/v4.36.1/src/transformers/trainer_callback.py#L564) when initializing `Trainer`, including when resuming from checkpoint. This means that if, for example, you never train your model for `early_stopping_patience`-many evaluation steps at once before stopping and resuming from checkpoint, early stopping will never happen.

An auxiliary issue is that even if you train your model for longer than `early_stopping_patience`-many evaluation steps, and training correctly stops, if you happen to then re-initiate training from a checkpoint, training will resume [even though the run ended with `self.control.should_training_stop == True`](https://github.com/huggingface/transformers/blob/f4f57f9dfa68948a383c352a900d588f63f6290a/src/transformers/trainer_callback.py#L154). This is because this variable is also not persisted to the `trainer_state.json` file when checkpointing. This issue was reported in #10290, but was never resolved before the issue was closed as stale.

To reproduce the main issue, simply initiate a training run and set `early_stopping_patience` to a value of your choice, then interrupt training before the run gets there. Reinitiate training with `resume_from_checkpoint=True`. Rinse and repeat until `best_metric` increases for `early_stopping_patience`-many evaluation calls.

To reproduce the auxiliary issue, don't interrupt your run until it stops due to early stopping. When it is complete, reinitiate training with `resume_from_checkpoint=True`.

### Expected behavior

Early stopping patience should work exactly the same when stopping and resuming runs from a checkpoint as when training continuously without interruption.

### PR DESCRIPTION
**Pull request by muellerzr**
# What does this PR do?

This PR builds a foundation for stateful callbacks inside `Trainer`. Right now these are isolated to the core callbacks that exist, but can be expanded upon as needed later.

Specifically a user must enable `restore_callback_states` in the `TrainingArguments` to enable this behavior.

## System Design:

To keep with not having to deal with pickles/interact with the `TrainerState` properly (where callback data should be stored, likely), we need a way to recreate the exact state of a `Callback`. To do so callbacks should implement a `save_state` function which will create a `dict` of `args` and `attributes` that should be set.

From this we can then recreate the constructor of an existing callback with these new options. (this is done in case there is logic in a callback's `__init__` that could be needed).

For example, here is the one for `EarlyStoppingCallback`:

```python
        def save(self) -> dict:
            return {
                "args": {
                    "early_stopping_patience": self.early_stopping_patience,
                    "early_stopping_threshold": self.early_stopping_threshold,
                },
                "attributes": {
                    "early_stopping_patience_counter": self.early_stopping_patience_counter,
                }
            }
```

## What about Callbacks that I forget to add back in?

As this relies on a similar **state**, I chose to keep aligned with how we deal with states/checkpointing in Accelerate, wherein users must recreate the exact same initial scenario when resuming training.

Or, in other words:

If you included the `EarlyStoppingCheckpoint` initially, but when you do `resume_from_checkpoint` you did not include that callback, we do not magically init and resume those states. Instead we will simply keep going and give you a nice warning message saying that things went amiss/some callback states weren't resumed.

## Limitations

As we're maintaining that these exist in the `TrainerState` (which again, makes sense for us to use this here), items should be JSON-serializable. 

The aim here is most likely not many callbacks actually *need* this, so only a few as needed over time can be added to this without much complexity.

Fixes https://github.com/huggingface/transformers/issues/28544

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@amyeroberts

### CODE REVIEW

**Comment by muellerzr** [comment]
@amyeroberts let me know if we need more docs or if I should rewrite the code a bit if it feels to complex/magical :)

File: tests/trainer/test_trainer_callback.py [hunk 1]
```diff
@@ -243,3 +247,76 @@ def test_event_flow(self):
                 callbacks=[MyTestTrainerCallback, MyTestTrainerCallback],
             )
             assert str(MyTestTrainerCallback) in warn_mock.call_args[0][0]
+
```
Comment (amyeroberts) [inline tests/trainer/test_trainer_callback.py:250]: Can we add a test/tests which pass in more than one callback (stateful and non-stateful) to make sure behaviour is as expected, and one which passes in two stateful callbacks and loads the correct state in.

**Review by amyeroberts — COMMENTED** [review]
Thanks for much for working on this - and for the detailed description in the PR! 

I understand the choice to match with accelerate patterns, but I think we might want to iterate on that. Looking at the behaviour in the tests I found it surprising, especially in the case when one can pass in a callback with default and it's previous state is loaded in: it seems to go against needing to fully specify the same training to resume. 

I'm also not sure on the loading of args vs attributes - it seems attributes we'd definitely want to store, but args we might want to be able to overwrite?

File: tests/trainer/test_trainer_callback.py [hunk 2]
```diff
@@ -243,3 +247,76 @@ def test_event_flow(self):
...
+            max_steps=2,
+        )
+        # Load it back in and verify values
+        checkpoint = os.path.join(self.output_dir, "checkpoint-2")
+        trainer.train(resume_from_checkpoint=checkpoint)
+        cb = [
+            callback for callback in trainer.callback_handler.callbacks if isinstance(callback, EarlyStoppingCallback)
+        ][0]
+        assert cb.early_stopping_patience == 5
+        assert cb.early_stopping_threshold == 0.2
```
[the remark this answers is not in the document]
Comment (muellerzr) [reply tests/trainer/test_trainer_callback.py:282]: So then the question becomes:

Should a user expect callbacks to magically appear if not specified, and if a `callback` is explicitly passed in then use its values? (Not the biggest fan, I'd rather users be explicit rather than implicit on what they want). 

Or should we do it this way.

The other option is set it in the `TrainingArguments` instead, something like `restore_callback_states_on_resume`

[the remark this answers is not in the document]
Comment (muellerzr) [reply tests/trainer/test_trainer_callback.py:282]: TBH, doing it in TrainingArguments might be preferred

File: tests/trainer/test_trainer_callback.py [hunk 3]
```diff
@@ -243,3 +247,76 @@ def test_event_flow(self):
...
+        # Create a new trainer with defaults
+        trainer = self.get_trainer(
+            max_steps=2,
+        )
+        # Load it back in and verify values
+        checkpoint = os.path.join(self.output_dir, "checkpoint-2")
+        # warning should be emitted for not-present callbacks
+        with patch("transformers.trainer.logger.warning") as warn_mock:
+            trainer.train(resume_from_checkpoint=checkpoint)
+            assert "EarlyStoppingCallback" in warn_mock.call_args[0][0]
```
[the remark this answers is not in the document]
Comment (muellerzr) [reply tests/trainer/test_trainer_callback.py:352]: Done, this has been solved by needing to explicitly enable it

File: src/transformers/trainer.py [hunk 4]
```diff
@@ -2711,6 +2724,38 @@ def opt_load_hook(mod, opt):
...
+                        else:
+                            # We remove the existing callback and add a new one
+                            self.callback_handler.remove_callback(callback)
+                            self.callback_handler.add_callback(new_callback)
+                logger.info("  Continuing training from checkpoint, restoring any callbacks that were passed in")
+            else:
+                not_found.append(stored_callback)
+        if len(not_found) > 0:
+            logger.warning(
+                f"  Checkpoint included callbacks not included in current configuration. Ignoring. ({', '.join(not_found)})"
```
Comment (amyeroberts) [inline src/transformers/trainer.py:2756]: nit

```suggestion
                f"Checkpoint included callbacks not included in current configuration. Ignoring. ({', '.join(not_found)})"
```

File: src/transformers/trainer.py [hunk 5]
```diff
@@ -2711,6 +2724,38 @@ def opt_load_hook(mod, opt):
...
+                        for attribute, value in attributes.items():
+                            setattr(new_callback, attribute, value)
+                        if isinstance(callback, TrainerControl):
+                            # Specifically for restoring the `control` state
+                            self.control = new_callback
+                        else:
+                            # We remove the existing callback and add a new one
+                            self.callback_handler.remove_callback(callback)
+                            self.callback_handler.add_callback(new_callback)
+                logger.info("  Continuing training from checkpoint, restoring any callbacks that were passed in")
```
Comment (amyeroberts) [inline src/transformers/trainer.py:2751]:
```suggestion
                logger.info("Continuing training from checkpoint, restoring any callbacks that were passed in")
```

File: src/transformers/trainer.py [hunk 6]
```diff
@@ -2711,6 +2724,38 @@ def opt_load_hook(mod, opt):
                     self.lr_scheduler.load_state_dict(torch.load(os.path.join(checkpoint, SCHEDULER_NAME)))
                 reissue_pt_warnings(caught_warnings)
 
+    def _load_callback_state(self):
+        """If callback states exist and were passed in, restore their states if enabled"""
+        if not self.args.restore_callback_states_from_checkpoint:
+            return
+        # Callback states are stored in stateful_callbacks
+        not_found = []
+        for stored_callback, data in self.state.stateful_callbacks.items():
```
Comment (amyeroberts) [inline src/transformers/trainer.py:2733]: Can we ever have more than one callback of the same type e.g. different metric loggers? Only concern with logic below is that we might load state from one class into another

**Review by amyeroberts — APPROVED** [review]
Thanks for iterating on this! I like this state control + warnings

File: src/transformers/trainer.py (hunk 6 above)
Comment (muellerzr) [reply src/transformers/trainer.py:2735]: Good call Amy, just in case we can, I've gone ahead and added logic in + a test to verify :)

**Comment by muellerzr** [comment]
@amyeroberts I'd like one more review please, just to make sure the multiple-versions logic makes sense to you!

File: tests/trainer/test_trainer_callback.py [hunk 7]
```diff
@@ -243,3 +261,156 @@ def test_event_flow(self):
...
+        checkpoint = os.path.join(self.output_dir, "checkpoint-2")
+        trainer.train(resume_from_checkpoint=checkpoint)
+        cbs = [
+            callback
+            for callback in trainer.callback_handler.callbacks
+            if isinstance(callback, MyTestExportableCallback)
+        ]
+        assert len(cbs) == 2
+        assert cbs[0].my_test_state == "first"
+        assert cbs[1].my_test_state == "second"
```
Comment (amyeroberts) [inline tests/trainer/test_trainer_callback.py:377]: Nice :)

File: src/transformers/trainer_callback.py [hunk 8]
```diff
@@ -103,10 +106,34 @@ class TrainerState:
...
+                    )
+                name = callback.__class__.__name__
+                if name in stateful_callbacks:
+                    # We can have multiple versions of the same callback
+                    # if so, we store them as a list of states to restore
+                    if not isinstance(stateful_callbacks[name], list):
+                        stateful_callbacks[name] = [stateful_callbacks[name]]
+                    stateful_callbacks[name].append(callback.state())
+                else:
+                    stateful_callbacks[name] = callback.state()
```
Comment (amyeroberts) [inline src/transformers/trainer_callback.py:135]: Is there a reason for not just always adding as a list directly. If we use a default dict we can just do 

```py
            # Saveable callbacks get stored as dict of kwargs
            stateful_callbacks = default_dict(list)
            for callback in self.stateful_callbacks:
                if not isinstance(callback, (ExportableState)):
                    raise TypeError(
                        f"All callbacks passed to be saved must inherit `ExportableState`, but received {type(callback)}"
                    )
                name = callback.__class__.__name__
                stateful_callbacks[name].append(callback.state())
```

**Review by amyeroberts — APPROVED** [review]
Thanks for iterating and for all the tests - looks great!

File: src/transformers/trainer_callback.py (hunk 8 above)
Comment (muellerzr) [reply src/transformers/trainer_callback.py:135]: Because that won't actually work 👀

Comment (muellerzr) [reply src/transformers/trainer_callback.py:136]: Python serialization does not like defaultdict, come to find out 😢

Comment (amyeroberts) [reply src/transformers/trainer_callback.py:136]: 🥲

**Comment by pedrobrs** [comment]
Hi,  it looks like the state of stateful callbacks are not saved to trainer_state.json on  _save_checkpoint method:

https://github.com/huggingface/transformers/blob/fe008d6ebea1f5770b740991daeefd9322fa434a/src/transformers/trainer.py#L2917-L2920
Only TrainerControl state is saved, instead of getting all ExportableState callbacks like here:
https://github.com/huggingface/transformers/blob/fe008d6ebea1f5770b740991daeefd9322fa434a/src/transformers/trainer.py#L674-L675

So it should be:
```
if self.args.should_save: 
    for cb in [cb for cb in self.callback_handler.callbacks  + [self.control] if isinstance(cb, ExportableState)]:
        self.state.stateful_callbacks[cb.__class__.__name__] == statefuf_cb.state()
    self.state.save_to_json(os.path.join(output_dir, TRAINER_STATE_NAME))    
```
Do you agree?
